sidechain/consensus/aura/src/peer_block_sync.rs

```diff
@@ -0,0 +1,193 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::block_production_suspension::{IsBlockProductionSuspended, SuspendBlockProduction};
+use core::marker::PhantomData;
+use its_consensus_common::{BlockImport, Error, Result};
+use its_primitives::{traits::SignedBlock as SignedSidechainBlockTrait, types::BlockHash};
+use log::*;
+use sp_runtime::traits::{Block as ParentchainBlockTrait, Header as ParentchainHeaderTrait};
+use std::{sync::Arc, vec::Vec};
+
+/// Trait for syncing sidechain blocks from a peer validateer.
+///
+/// This entails importing blocks and detecting if we're out of date with our blocks, in which
+/// case we fetch the missing blocks from a peer.
+pub trait SyncBlockFromPeer<ParentchainHeader, SignedSidechainBlock>
+where
+	ParentchainHeader: ParentchainHeaderTrait,
+	SignedSidechainBlock: SignedSidechainBlockTrait,
+{
+	fn attempt_block_sync(
+		&self,
+		sidechain_block: SignedSidechainBlock,
+		last_imported_parentchain_header: &ParentchainHeader,
+	) -> Result<()>;
+}
+
+/// Sidechain peer block sync implementation.
+pub struct PeerBlockSync<
+	ParentchainBlock,
+	SignedSidechainBlock,
+	BlockImporter,
+	BlockProductionSuspender,
+> {
+	importer: Arc<BlockImporter>,
+	block_production_suspender: Arc<BlockProductionSuspender>,
+	_phantom: PhantomData<(ParentchainBlock, SignedSidechainBlock)>,
+}
+
+impl<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+	PeerBlockSync<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+where
+	ParentchainBlock: ParentchainBlockTrait,
+	SignedSidechainBlock: SignedSidechainBlockTrait,
+	BlockImporter: BlockImport<ParentchainBlock, SignedSidechainBlock>,
+	BlockProductionSuspender: SuspendBlockProduction + IsBlockProductionSuspended,
+{
+	pub fn new(
+		importer: Arc<BlockImporter>,
+		block_production_suspender: Arc<BlockProductionSuspender>,
+	) -> Self {
+		PeerBlockSync { importer, block_production_suspender, _phantom: Default::default() }
+	}
+
+	fn sync_blocks_from_peer(
```

Comment:
I think the names are getting a little confusing.

We have a `PeerBlockSync` which syncs sidechain blocks from a peer, no matter if they have been fetched or gossiped, I can agree to that. 

But this function has effectively the same name, but only syncs blocks via fetching. Maybe rename to
`fetch_and_sync_blocks_from_peer` ?

Reply:
Would `fetch_and_import_blocks_from_peer` be acceptable too? Because that's exactly what it does

Reply:
Perfectly so 👍